### PR TITLE
Rewrite Inventory Handling

### LINKED ISSUE #3810: /sell hand with the same item in offhand results in error
### Information

**Full output of `/ess version`:**
```
Server version: 1.16.4-R0.1-SNAPSHOT git-Paper-313 (MC: 1.16.4)
EssentialsX version: 2.18.2.0
LuckPerms version: 5.1.93
Vault version 1.7.3-b131
EssentialsXProtect version: 2.18.2.0
EssentialsXGeoIP version: 2.18.2.0
EssentialsXChat version: 2.18.2.0
EssentialsXAntiBuild version: 2.18.2.0
EssentialsXSpawn version: 2.18.2.0
```

**Server startup log:**
[latest.log](https://paste.gg/p/anonymous/a3d9d7df7ed1447c9103d6ee83dd61bf)

**EssentialsX config:**
[config.yml.txt](https://paste.gg/p/anonymous/0dd9f4b661ea4637ad03eef6a94922c5)

### Details

**Description:**  
using /sell hand when the same type of item is in the offhand results in the following error: `Error: Trying to remove more items than are available`

**Steps to reproduce:**  
Use /sell hand on an item with the same item in the offhand

**Expected behavior:**  
Items should be sold as normal

### LINKED ISSUE #4248: EssentialsX /clear Not Working As Expected
### Type of bug

Other unexpected behaviour

### `/ess version` output

Server version: 1.16.5-R0.1-SNAPSHOT 2991-Spigot (MC: 1.16.5)
EssentialsX version: 2.18.2.0
EssentialsXChat version: 2.18.2.0
LuckPerms version: 5.3.47

### Server startup log

https://paste.gg/p/anonymous/0312d1109ab0492bbc52b0978fa64021

### EssentialsX config files

https://paste.gg/p/anonymous/0312d1109ab0492bbc52b0978fa64021

### Error log (if applicable)

https://paste.gg/p/anonymous/0312d1109ab0492bbc52b0978fa64021

### Bug description

When executing ` clear PLAYERNAMEHERE COBBLESTONE 64 ` in console, if a player has any number of cobblestone in their off-hand slot, it is not removed.

I expected to be able to remove any specified number of an item using that command even if it was in their off-hand slot or in their inventory.

Pls fix. ;o;

### Steps to reproduce

Execute ` clear PLAYERNAMEHERE COBBLESTONE 64 ` in console while a player has any number of cobblestone in their off-hand slot and it will not be removed.

### Expected behaviour

I expected to be able to remove any specified number of an item using that command even if it was in their off-hand slot or in their inventory.

### Actual behaviour

Only removes the specified number of items from the inventory, not their armor or off-hand slot as well.

### PR DESCRIPTION
**Pull request by JRoy**
there are a LOT of weird inventory stuff going on in EssX.

Fixes #3810
Fixes #4248

maybe more inventory issues that I'm unable to find?

### CODE REVIEW

**Comment by JRoy** [comment]
ready for review, needs testing with all commands/features that touch items.

File: Essentials/src/main/java/com/earth2me/essentials/Kit.java [hunk 1]
```diff
@@ -214,57 +213,34 @@ public boolean expandItems(final User user, final List<String> items) throws Exc
...
+            final ItemStack[] itemArray = itemList.toArray(new ItemStack[0]);
+            if (!isDropItemsIfFull && !Inventories.hasSpace(user.getBase(), maxStackSize, autoEquip, itemArray)) {
+                user.sendMessage(tl("kitInvFullNoDrop"));
+                return false;
+            }
+
+            final Map<Integer, ItemStack> leftover = Inventories.addItem(user.getBase(), maxStackSize, autoEquip, itemArray);
+            if (!isDropItemsIfFull && !leftover.isEmpty()) {
+                // Inventories#hasSpace should prevent this state from EVER being reached; If it does, something has gone terribly wrong, and we should just give up and hope people report it :(
+                throw new IllegalStateException(leftover.toString());
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/Kit.java]: Can we make a more user-friendly error message for this instead of just dumping the items

File: Essentials/src/main/java/com/earth2me/essentials/Kit.java [hunk 2]
```diff
@@ -174,6 +172,7 @@ public boolean expandItems(final User user, final List<String> items) throws Exc
             final List<String> commandQueue = new ArrayList<>();
             final List<String> moneyQueue = new ArrayList<>();
             final String currencySymbol = ess.getSettings().getCurrencySymbol().isEmpty() ? "$" : ess.getSettings().getCurrencySymbol();
+
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/Kit.java]: unnecessary diff

File: Essentials/src/main/java/com/earth2me/essentials/craftbukkit/Inventories.java [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+
+public final class Inventories {
+    private static final int HELM_SLOT = 39;
+    private static final int CHEST_SLOT = 38;
+    private static final int LEG_SLOT = 37;
+    private static final int BOOT_SLOT = 36;
+    private static final boolean IS_OFFHAND = VersionUtil.getServerBukkitVersion().isHigherThanOrEqualTo(VersionUtil.v1_9_R01);
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/craftbukkit/Inventories.java]: for clarity
```suggestion
    private static final boolean HAS_OFFHAND = VersionUtil.getServerBukkitVersion().isHigherThanOrEqualTo(VersionUtil.v1_9_R01);
```

File: Essentials/src/main/java/com/earth2me/essentials/craftbukkit/Inventories.java [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+                }
+            }
+        }
+        return false;
+    }
+
+    public static boolean hasSpace(final Player player, final int maxStack, final boolean includeArmor, ItemStack... items) {
+        items = normalizeItems(cloneItems(items));
+        final InventoryData inventoryData = parseInventoryData(player.getInventory(), items, maxStack, includeArmor);
+        System.out.println(inventoryData.getEmptySlots());
```
Comment (pop4959) [inline Essentials/src/main/java/com/earth2me/essentials/craftbukkit/Inventories.java]: just a reminder to remove sysout

**Review by pop4959 — COMMENTED** [review]
Just a few notes on initial look-through

**Comment by pop4959** [comment]
Tested commands: `/book, /clearinventory, /condense, /createkit, /enchant, /hat, /item, /itemlore, /itemname, /kit, /powertool, /repair, /sell, /setworth, /skull, /spawnnmob, /unlimited, /worth`

Didn't test keep inventory policy or trade/enchant signs yet.

Looks good so far though, didn't find any issues with any commands.

**Review by pop4959 — APPROVED** [review]

**Comment by pop4959** [comment]
Still haven't tested the things above, but assuming those are ok this PR looks good to me

File: Essentials/src/main/java/com/earth2me/essentials/User.java [hunk 5]
```diff
@@ -1137,12 +1137,7 @@ public void setConfirmingClearCommand(final String command) {
      * Returns the {@link ItemStack} in the main hand or off-hand. If the main hand is empty then the offhand item is returned - also nullable.
      */
     public ItemStack getItemInHand() {
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/User.java:1143]: Do we want to mark as `@Deprecated` or just leave this be?

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java [hunk 6]
```diff
@@ -154,6 +157,10 @@ public void run(final Server server, final CommandSource sender, final String co
                 runUserMap(sender, args);
                 break;
 
+            case "itemtest":
+                runItemTest(server, sender, commandLabel, args);
```
Comment (mdcfe) [inline Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java:161]: Just checking whether you meant to leave this in here?

**Review by mdcfe — COMMENTED** [review]
Looks good, just a couple of things to note.

File: Essentials/src/main/java/com/earth2me/essentials/User.java (hunk 5 above)
Comment (JRoy) [reply Essentials/src/main/java/com/earth2me/essentials/User.java:1143]: Leave be, this still has valid uses

File: Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java (hunk 6 above)
Comment (JRoy) [reply Essentials/src/main/java/com/earth2me/essentials/commands/Commandessentials.java:161]: Yep

**Review by mdcfe — APPROVED** [review]
